Create a custom web editor using Monaco and Language Server Protocol (LSP)

cjayashantha opened the pull request.

romanutti commented on drafts/web-editor-monaco-with-language-server-support.md:
```diff
@@ -0,0 +1,971 @@
+---
+title: Create a custom web editor using Monaco and Language Server Protocol (LSP)
+domain: software-engineering-corner.hashnode.dev
+tags: Web Editor, Monaco, Monaco Language Client, Language Server, LSP
```
Regarding the tags, in our readme its recommented to `Use tags which are listed here https://github.com/Hashnode/support/blob/main/misc/tags.json (use the slug) or ensure they exist via the Hashnode search (select "tags"). If you do it wrong, Hashnode may fail to import the article.`

romanutti commented on drafts/web-editor-monaco-with-language-server-support.md:
```diff
@@ -0,0 +1,971 @@
+domain: software-engineering-corner.hashnode.dev
+tags: Web Editor, Monaco, Monaco Language Client, Language Server, LSP
+cover: https://cdn.hashnode.com/res/hashnode/image/upload/v1719718605895/bv2LfyFWS.jpg?auto=format
+publishAs: cjayashantha, NipunaMarcusZuhlke
```
Hope multiple authors work 🤞🏼

romanutti commented on drafts/web-editor-monaco-with-language-server-support.md:
```diff
@@ -0,0 +1,971 @@
+
+# Create a custom web editor using Monaco and Language Server Protocol (LSP)
+
+If you are developing a domain-specific language or custom language, chances are high that current editors might not support features such as syntax highlighting, diagnostics, and code autocompletion. To enable these features, you need a Language Server Protocol (LSP) and sometimes a custom editor. In this article, we will discuss how to develop a web editor and connect it to a language server us ...
```
Here you use `autocompletion`, while you use `auto-completion` everywhere else in the article. Probably would try to be consistent

romanutti replied:
Maybe from a readers perspective it would easier to follow if you show the screenshort/gif that we should at the end already here. Then it would be super clear what we are aiming for through the articles

romanutti replied:
Would still vote for showing the [final screen](https://github.com/Zuehlke/software-engineering-corner/pull/31/files#diff-5ddfe36215164246e909db4ba17080865f69c8de7c209b9af503d93106cf2f8eR415) upfront in the first article to show what we want to achieve 😄

romanutti commented on drafts/web-editor-monaco-with-language-server-support.md:
```diff
@@ -0,0 +1,971 @@
+
+For this example, we will develop our own mock language server fronted with websocket. Full implementation available at below location
+
+https://github.com/NipunaMarcus/hellols/tree/websocket-launcher
```
You could use markdown syntax for links:
`[link text](https://link-url-here.org)`

This is then displayed as [link text](https://link-url-here.org)

romanutti commented on drafts/web-editor-monaco-with-language-server-support.md:
```diff
@@ -0,0 +1,971 @@
+![Architecture](https://cdn.hashnode.com/res/hashnode/image/upload/v1717742534914/GzLBzavNE.png?auto=format)
+
+## Creating the Language Server
+Full Backend implementation can be found at the below GitHub repo
```
Again, you could directly link the repo using the syntax described in https://github.com/Zuehlke/software-engineering-corner/pull/31/files#r1664745293

romanutti commented on drafts/web-editor-monaco-with-language-server-support.md:
```diff
@@ -0,0 +1,971 @@
+### Step 1: Language Server Implementation
+For the Language server implementation here I'm using [LSP4J](https://github.com/eclipse-lsp4j/lsp4j). As you can see there are three classes available under the language-server module. So these three classes are added to implement the three main interfaces provided by LSP4J as to the breakdown given in the LSP specification to support general cases, language features, and workspace management.
+
+* HelloLanguageServer.java — This class implements the interface available in the LSP4J called LanguageServer which contains the general functionality of the language server such as initializing the language server, shutting down the language server and so on … Also if the LS needs to publish the diagnostics(compilation error and semantic errors) back to the Client (in our case VSCode plugin) LS  ...
```
Missing whitespace at `diagnostics (compi...`

romanutti replied:
Maybe you could introduce the abbrevation Langugae Serve (LS) before using it

romanutti commented on drafts/web-editor-monaco-with-language-server-support.md:
```diff
@@ -0,0 +1,971 @@
+
+![web-editor](https://cdn.hashnode.com/res/hashnode/image/upload/v1717756040911/IFeF6apmJ.png?auto=format)
+
+## Creating Web Editor using React
```
As this article is already quite a long read you could ask yourself if you want to split it to two (lsp & angular / react) or even three articles (lsp / angular / react)

Cybertron01Z replied:
i like this idea, in hashnode we can do series

romanutti commented on drafts/web-editor-monaco-with-language-server-support.md:
```diff
@@ -0,0 +1,971 @@
+
+import java.util.concurrent.CompletableFuture;
+
+public class HelloLanguageServer implements LanguageServer, LanguageClientAware {
```
I was a bit confused of the `Hello`-prefix of those classes - would it be worth naming them more after their intetions, e.g. `LanguageServer`, `LanguageFeaturesService` , `WorkspaceManagementService`

romanutti commented on drafts/web-editor-monaco-with-language-server-support.md:
```diff
@@ -0,0 +1,971 @@
+
+So if you look at the implementation I’m just creating a `CompletionItem` and filling it with what type of completion item is this and what is the text to be inserted and description and label of the completion item.
+
+If you have a language AST or Source processor(string processor), this is the place where you can put your logic to filter out what to be provided as completion based on the line, column and the file which are provided as a parameter to the `completion()` method as `CompletionParams`.
```
What is a language `AST`?

Cybertron01Z replied:
abstract syntax tree i guess

romanutti commented on drafts/web-editor-monaco-with-language-server-support.md:
````diff
@@ -0,0 +1,971 @@
+```
+Here is the final look.
+
+![React Web Editor](https://cdn.hashnode.com/res/hashnode/image/upload/v1719713207751/GDUJ_Rq1a.jpeg?auto=format)
````
Definitely a nice to have but does hashnode also support gifs?

Cybertron01Z replied:
it should yes and it would also support a codesandbox, in case this would run there. Would be pretty cool to have a working example here with all the code

romanutti commented:
Good work guys! In-depth and well explained 👏🏼 

Most relevant would probably be the question if you want to split it into multiple articles as mentioned here: https://github.com/Zuehlke/software-engineering-corner/pull/31/files#r1664768250

NipunaMarcusZuhlke commented:
@romanutti article now broken down to three articles ... To make it a series we have to do it in the UI? or is there any other way?

Cybertron01Z commented on drafts/develop-a-web-editor-with-react-and-monaco-with-language-server-support.md:
```diff
@@ -0,0 +1,334 @@
+title: Develop a Web Editor for your DSL using React and Monaco Editor library with Language Server support
+domain: software-engineering-corner.hashnode.dev
+tags: Web Editor, Monaco, Monaco Language Client, Language Server, LSP
+cover: https://cdn.hashnode.com/res/hashnode/image/upload/v1720940174478/yjy4aJtIT.gif?auto=format
```
nitpick:
would it be possible to make this dark mode so it fits better into our blog?

romanutti commented on drafts/develop-your-own-language-server.md:
```diff
@@ -0,0 +1,322 @@
+
+# Develop your own Language Server
+
+If you have ever worked with an IDE for programming languages like Java or C#, you have probably experience that how easy it is to write program with auto-completion feature ( some might know it as intellisense ) provided by the IDE which help you to type code faster and select suitable functions to complete your task, or find the usage of the particular function, or go to the implementation of a ...
```
`(some might know it as intellisense)`

romanutti commented on drafts/develop-your-own-language-server.md:
```diff
@@ -0,0 +1,322 @@
+If you have ever worked with an IDE for programming languages like Java or C#, you have probably experience that how easy it is to write program with auto-completion feature ( some might know it as intellisense ) provided by the IDE which help you to type code faster and select suitable functions to complete your task, or find the usage of the particular function, or go to the implementation of a ...
+
+## So where is this so-called language server fit in?
+If you are a developer (Obviously) you might have a favorite IDE that you always use. And it is really easy if you can develop software from different languages on top of the same IDE that you like. So various developers choose IDEs that they like and stick to it. Because of that language developers always trying to add language feature support for all most all the IDEs and that is hard because e ...
```
`(obviously)`

romanutti replied:
`(architecturally)`

romanutti commented on drafts/develop-a-web-editor-with-angular-and-monaco-with-language-server-support.md:
```diff
@@ -0,0 +1,425 @@
+- Multiple language support
+
+### What is Language Server Protocol (LSP) ?
+LSP is a protocol used to provide language-specific features in a language-agnostic way. It decouples the editor from the language-specific logic, allowing you to support various languages with minimal effort. LSP offers:
```
You could also add the link to the last article here

cjayashantha replied:
Added in line 15.

romanutti commented on drafts/develop-a-web-editor-with-angular-and-monaco-with-language-server-support.md:
````diff
@@ -0,0 +1,425 @@
+ng new web-editor
+```
+
+- Navigate to generated project folder
````
Missing empty line - this leads to a intended code section

tispBe commented:
@cjayashantha what is the state of the article and revisions? are you still working on it or can we merge it?

cjayashantha commented:
> @cjayashantha what is the state of the article and revisions? are you still working on it or can we merge it?

@tispBe Yes,  we can merge this.  we have addressed the review comments.

tispBe commented on drafts/develop-a-web-editor-with-angular-and-monaco-with-language-server-support.md:
```diff
@@ -0,0 +1,489 @@
+cover: https://cdn.hashnode.com/res/hashnode/image/upload/v1719718605895/bv2LfyFWS.jpg?auto=format
+publishAs: cjayashantha
+seriesSlug: editor-with-ls-support
+hideFromHashnodeCommunity: false
```
```suggestion
hideFromHashnodeCommunity: false
saveAsDraft: true
```
